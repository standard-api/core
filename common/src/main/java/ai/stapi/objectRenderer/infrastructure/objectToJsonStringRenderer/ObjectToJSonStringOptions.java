package ai.stapi.objectRenderer.infrastructure.objectToJsonStringRenderer;


import ai.stapi.objectRenderer.model.RendererOptions;
import java.util.Arrays;
import java.util.List;

public class ObjectToJSonStringOptions implements RendererOptions {

  private final List<RenderFeature> features;

  public ObjectToJSonStringOptions(RenderFeature... renderFeatures) {
    this.features = Arrays.stream(renderFeatures).toList();
  }

  public ObjectToJSonStringOptions(List<RenderFeature> renderFeatures) {
    this.features = renderFeatures;
  }

  public List<RenderFeature> getFeatures() {
    return features;
  }

  public enum RenderFeature {
    SORT_FIELDS,
    HIDE_CREATED_AT,
    HIDE_DISPATCHED_AT,
    HIDE_IDS,
    HIDE_KEY_HASHCODE, RENDER_GETTERS
  }
}