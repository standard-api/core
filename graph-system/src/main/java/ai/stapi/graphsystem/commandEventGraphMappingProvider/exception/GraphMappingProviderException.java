package ai.stapi.graphsystem.commandEventGraphMappingProvider.exception;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import ai.stapi.graphsystem.commandEventGraphMappingProvider.specific.SpecificCommandEventGraphMappingProvider;
import java.util.List;
import org.apache.commons.lang3.StringUtils;

public class GraphMappingProviderException extends RuntimeException {

  private GraphMappingProviderException(String message) {
    super(message);
  }

  public static GraphMappingProviderException becauseThereIsNoSupportingSpecificGraphMappingProvider(
      String deserializationType) {
    return new GraphMappingProviderException(
        String.format(
            "\nThere is no supporting %s"
                + " for object of type:\n%s",
            SpecificCommandEventGraphMappingProvider.class.getSimpleName(),
            deserializationType
        )
    );
  }

  public static GraphMappingProviderException becauseThereAreMoreThanOneSpecificGraphMappingProviders(
      Class<?> classType,
      List<SpecificCommandEventGraphMappingProvider> listOfSupportingProviders
  ) {
    return new GraphMappingProviderException(
        "Object of type '"
            + classType.getSimpleName()
            + "' is supported by multiple providers ("
            + StringUtils.join(listOfSupportingProviders, ", ")
            + ") and that is not allowed."
    );
  }

  public static GraphMappingProviderException becauseCommandDoesNotContainGivenField(
      Class<? extends AbstractCommand> classType,
      String fieldName
  ) {
    return new GraphMappingProviderException(
        "Command of type '" + classType.getSimpleName() + "' does not contain field with name '"
            + fieldName + "'."
    );
  }
}
