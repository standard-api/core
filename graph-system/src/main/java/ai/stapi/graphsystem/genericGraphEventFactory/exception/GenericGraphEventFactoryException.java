package ai.stapi.graphsystem.genericGraphEventFactory.exception;

import ai.stapi.graphsystem.messaging.event.AggregateGraphUpdatedEvent;

public class GenericGraphEventFactoryException extends RuntimeException {

  private GenericGraphEventFactoryException(String message) {
    super(message);
  }

  public static GenericGraphEventFactoryException becauseNoSupportingSpecificFactoriesForGivenEvent(
      Class<? extends AggregateGraphUpdatedEvent> event
  ) {
    return new GenericGraphEventFactoryException(
        "There are no supporting factories for event '" + event.getSimpleName() + "'."
    );
  }

  public static GenericGraphEventFactoryException becauseMoreThanOneSpecificFactoriesForGivenCommand(
      Class<? extends AggregateGraphUpdatedEvent> event) {
    return new GenericGraphEventFactoryException(
        "There are multiple supporting factories for event '"
            + event.getSimpleName()
            + "' and that is not allowed."
    );
  }
}
