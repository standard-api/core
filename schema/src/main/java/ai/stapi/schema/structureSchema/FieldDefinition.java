package ai.stapi.schema.structureSchema;

import ai.stapi.serialization.AbstractSerializableObject;
import java.util.Comparator;
import java.util.List;

public class FieldDefinition extends AbstractSerializableObject {

  public static final String SERIALIZATION_TYPE = "FieldDefinition";
  public static final String LIST_STRUCTURE_TYPE = "list";
  public static final String LEAF_STRUCTURE_TYPE = "leaf";
  private String name;
  private Integer min;
  private String max;
  private String description;
  private List<FieldType> types;

  private String parentDefinitionType;

  private FieldDefinition() {
    super(SERIALIZATION_TYPE);
  }

  public FieldDefinition(
      String name,
      Integer min,
      String max,
      String description,
      List<FieldType> types,
      String parentDefinitionType
  ) {
    super(SERIALIZATION_TYPE);
    this.name = name;
    this.min = min;
    this.max = max;
    this.description = description;
    this.types = types.stream()
        .sorted(Comparator.comparing(FieldType::getType))
        .toList();
    this.parentDefinitionType = parentDefinitionType;
  }

  public Integer getMin() {
    return min;
  }

  public String getMax() {
    return max;
  }

  public Float getFloatMax() {
    if (this.max.equals("*")) {
      return Float.POSITIVE_INFINITY;
    }
    return Float.parseFloat(this.max);
  }

  public List<FieldType> getTypes() {
    return types;
  }

  public String getStructureType() {
    return isList() ? LIST_STRUCTURE_TYPE : LEAF_STRUCTURE_TYPE;
  }

  public boolean isList() {
    try {
      var value = Integer.parseInt(this.max);
      return value > 1;
    } catch (NumberFormatException ignored) {

    }
    return this.max.equals("*");
  }

  public boolean isRequired() {
    return this.min > 0;
  }

  public boolean isUnionType() {
    return this.types.size() > 1;
  }

  public boolean isAnyType() {
    return this.types.isEmpty();
  }

  public String getDescription() {
    return description;
  }

  public String getName() {
    return name;
  }

  public String getParentDefinitionType() {
    return parentDefinitionType;
  }
}
