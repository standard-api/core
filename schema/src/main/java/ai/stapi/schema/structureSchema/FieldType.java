package ai.stapi.schema.structureSchema;

import org.apache.commons.lang3.StringUtils;

public class FieldType {

  private final String type;
  private final String originalType;
  private final FieldTypeGroup typeGroup;

  private final boolean isContentReferenced;

  public FieldType(
      String type,
      String originalType
  ) {
    this(type, originalType, FieldTypeGroup.STANDARD);
  }

  public FieldType(
      String type,
      String originalType,
      FieldTypeGroup fieldTypeGroup
  ) {
    this(type, originalType, fieldTypeGroup, false);
  }

  private FieldType(
      String type,
      String originalType,
      FieldTypeGroup fieldTypeGroup,
      boolean isContentReferenced
  ) {
    this.type = type;
    this.originalType = originalType;
    this.typeGroup = fieldTypeGroup;
    this.isContentReferenced = isContentReferenced;
  }

  public static FieldType asReferenceType(String type) {
    return new FieldType(type, "Reference", FieldTypeGroup.REFERENCE);
  }

  public static FieldType asContentReferenced(String type, String originalName) {
    return new FieldType(
        type,
        originalName,
        FieldTypeGroup.ANONYMOUS,
        true
    );
  }

  public static FieldType asAnonymousType(String type, String originalName) {
    if (!originalName.equals("Element") && !originalName.equals("BackboneElement")) {
      throw new RuntimeException(
          "Anonymous type's original name must be either Element or BackboneElement");
    }
    return new FieldType(type, originalName, FieldTypeGroup.ANONYMOUS);
  }

  public static FieldType asBoxedType(String type) {
    String typeName;
    String originalName;
    if (type.startsWith("Boxed")) {
      typeName = type;
      originalName = type.replace("Boxed", "").toLowerCase();
    } else {
      typeName = "Boxed" + StringUtils.capitalize(type);
      originalName = type;
    }
    return new FieldType(typeName, originalName, FieldTypeGroup.BOXED);
  }

  public static FieldType asPlainType(String type) {
    return new FieldType(type, type);
  }

  public String getType() {
    return type;
  }

  public String getOriginalType() {
    return originalType;
  }

  public FieldTypeGroup getTypeGroup() {
    return typeGroup;
  }

  public boolean isReference() {
    return this.originalType.equals("Reference");
  }

  public boolean isAnonymous() {
    return this.originalType.equals("Element") || this.originalType.equals("BackboneElement");
  }

  public boolean isBoxed() {
    return this.type.startsWith("Boxed") && !this.type.equals(this.originalType);
  }

  public boolean isPrimitiveType() {
    return this.type.equals(this.originalType) && Character.isLowerCase(this.type.charAt(0));
  }

  public boolean isContentReferenced() {
    return isContentReferenced;
  }
}
