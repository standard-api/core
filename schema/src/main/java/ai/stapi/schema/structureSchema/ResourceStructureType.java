package ai.stapi.schema.structureSchema;

import java.util.Map;

public class ResourceStructureType extends ComplexStructureType {

  public static final String SERIALIZATION_TYPE = "ResourceStructureType";
  public static final String KIND = AbstractStructureType.RESOURCE;

  private ResourceStructureType() {
    super();
  }

  public ResourceStructureType(
      String definitionType,
      Map<String, FieldDefinition> fields,
      String description,
      String parent,
      boolean isAbstract
  ) {
    super(
        SERIALIZATION_TYPE,
        AbstractStructureType.RESOURCE,
        definitionType,
        fields,
        description,
        parent,
        isAbstract
    );
  }

  @Override
  public ResourceStructureType copyWithNewFields(Map<String, FieldDefinition> newFields) {
    return new ResourceStructureType(
        definitionType,
        newFields,
        description,
        parent,
        isAbstract
    );
  }
}
