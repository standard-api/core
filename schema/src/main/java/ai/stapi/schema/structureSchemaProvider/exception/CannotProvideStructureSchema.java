package ai.stapi.schema.structureSchemaProvider.exception;

import ai.stapi.schema.structureSchemaMapper.UnresolvableType;
import java.util.List;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

public class CannotProvideStructureSchema extends Exception {

  private final String missingSerializationType;

  public CannotProvideStructureSchema(
      String missingSerializationType,
      List<UnresolvableType> unresolvableTypes
  ) {
    super(createMessage(missingSerializationType, unresolvableTypes).toString());
    this.missingSerializationType = missingSerializationType;
  }

  @NotNull
  private static StringBuilder createMessage(
      String missingSerializationType,
      List<UnresolvableType> unresolvableTypes
  ) {
    var errorMessage = new StringBuilder()
        .append("Structure schema for serialization type '")
        .append(missingSerializationType)
        .append("' not found.")
        .append(System.lineSeparator())
        .append("Current pending unresolvable types: ");
    unresolvableTypes.forEach(unresolvable -> {
      errorMessage
          .append(System.lineSeparator())
          .append("Failed type: '")
          .append(unresolvable.structureDefinitionData().getId())
          .append("', ")
          .append("missing dependencies: [")
          .append(StringUtils.join(unresolvable.missingDependencies(), ", "))
          .append("]");
    });
    return errorMessage;
  }

  public String getMissingSerializationType() {
    return missingSerializationType;
  }
}
