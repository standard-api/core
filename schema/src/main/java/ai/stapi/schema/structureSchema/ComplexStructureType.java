package ai.stapi.schema.structureSchema;

import ai.stapi.schema.structureSchema.exception.FieldDoesNotExist;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.stream.Collectors;

public class ComplexStructureType extends AbstractStructureType {

  public static final String SERIALIZATION_TYPE = "ComplexStructureType";
  public static final String KIND = AbstractStructureType.COMPLEX_TYPE;

  private Map<String, FieldDefinition> fields;

  protected ComplexStructureType() {
    super();
  }

  protected ComplexStructureType(
      String serializationType,
      String kind,
      String definitionType,
      Map<String, FieldDefinition> fields,
      String description,
      String parent,
      boolean isAbstract
  ) {
    super(serializationType, kind, definitionType, description, isAbstract, parent);
    this.fields = new HashMap<>(fields);
  }

  public ComplexStructureType(
      String definitionType,
      Map<String, FieldDefinition> fields,
      String description,
      String parent,
      boolean isAbstract

  ) {
    super(
        SERIALIZATION_TYPE,
        AbstractStructureType.COMPLEX_TYPE,
        definitionType,
        description,
        isAbstract,
        parent
    );
    this.fields = new HashMap<>(fields);
  }

  public Map<String, FieldDefinition> getAllFields() {
    return fields;
  }

  public Map<String, FieldDefinition> getOwnFields() {
    return fields.entrySet().stream()
        .filter(entry -> entry.getValue().getParentDefinitionType().equals(this.definitionType))
        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
  }

  public FieldDefinition getField(String fieldName) {
    if (!this.fields.containsKey(fieldName)) {
      throw new FieldDoesNotExist(
          String.format(
              "Field '%s' does not exist on '%s'.",
              fieldName,
              this.getDefinitionType()
          )
      );
    }
    return this.fields.get(fieldName);
  }

  public boolean hasField(String fieldName) {
    return this.fields.containsKey(fieldName);
  }

  public ComplexStructureType copyWithNewFields(Map<String, FieldDefinition> newFields) {
    return new ComplexStructureType(
        definitionType,
        newFields,
        description,
        parent,
        isAbstract
    );
  }
}
