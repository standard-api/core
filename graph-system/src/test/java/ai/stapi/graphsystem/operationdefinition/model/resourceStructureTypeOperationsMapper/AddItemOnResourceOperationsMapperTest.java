package ai.stapi.graphsystem.operationdefinition.model.resourceStructureTypeOperationsMapper;

import ai.stapi.objectRenderer.infrastructure.objectToJsonStringRenderer.ObjectToJSonStringOptions.RenderFeature;
import ai.stapi.schema.structureSchema.ComplexStructureType;
import ai.stapi.schema.structureSchema.FieldDefinition;
import ai.stapi.schema.structureSchema.FieldType;
import ai.stapi.schema.structureSchema.ResourceStructureType;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaProvider;
import ai.stapi.schema.structureSchemaProvider.exception.CannotProvideStructureSchema;
import ai.stapi.test.SystemSchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(AddItemOnResourceOperationsMapperTestDefinitionsLoader.SCOPE)
class AddItemOnResourceOperationsMapperTest extends SystemSchemaIntegrationTestCase {

  @Autowired
  private AddItemOnResourceOperationsMapper mapper;

  @Autowired
  private StructureSchemaProvider schemaProvider;

  @Test
  void itCanMapSimpleStructure() throws CannotProvideStructureSchema {
    var input = this.schemaProvider.provideSpecific("ExampleTestResource");
    var actual = this.mapper.map((ResourceStructureType) input);
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapDeeperFieldToAdd() {
    var actual = this.mapper.map(
        new ResourceStructureType(
            "ExampleResourceType",
            Map.of(
                "exampleElementField", new FieldDefinition(
                    "exampleElementField",
                    1,
                    "1",
                    "Example Element field on Example Resource Type",
                    List.of(
                        new FieldType("Timing", "Timing")
                    ),
                    "ExampleResourceType"
                )
            ),
            "This is a example resource type.",
            null,
            false
        )
    );
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapResourceWithUnionField() {
    var actual = this.mapper.map(
        new ResourceStructureType(
            "ExampleResourceType",
            Map.of(
                "exampleUnionField", new FieldDefinition(
                    "exampleUnionField",
                    1,
                    "1",
                    "Example Union field on Example Resource Type",
                    List.of(
                        new FieldType("BoxedDecimal", "decimal"),
                        new FieldType("StructureDefinition", "Reference"),
                        new FieldType("Timing", "Timing"),
                        new FieldType("SameFieldComplex1", "SameFieldComplex1"),
                        new FieldType("SameFieldComplex2", "SameFieldComplex2")
                    ),
                    "ExampleResourceType"
                )
            ),
            "This is a example resource type.",
            null,
            false
        )
    );
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapResourceWithUnionListField() {
    var actual = this.mapper.map(
        new ResourceStructureType(
            "ExampleResourceType",
            Map.of(
                "exampleUnionField", new FieldDefinition(
                    "exampleUnionField",
                    1,
                    "*",
                    "Example Union field on Example Resource Type",
                    List.of(
                        new FieldType("BoxedDecimal", "decimal"),
                        new FieldType("StructureDefinition", "Reference"),
                        new FieldType("Timing", "Timing"),
                        new FieldType("SameFieldComplex1", "SameFieldComplex1"),
                        new FieldType("SameFieldComplex2", "SameFieldComplex2")
                    ),
                    "ExampleResourceType"
                )
            ),
            "This is a example resource type.",
            null,
            false
        )
    );
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapOperationDefinition() throws CannotProvideStructureSchema {
    var input = this.schemaProvider.provideSpecific("OperationDefinition");
    var actual = this.mapper.map((ResourceStructureType) input);
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapNewFieldOnResource() throws CannotProvideStructureSchema {
    var modifiedResource = (ComplexStructureType) this.schemaProvider.provideSpecific(
        "AggregateDefinition"
    );
    var actual = this.mapper.mapNewFields(
        modifiedResource,
        List.of("command")
    );
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapNewInheritedFieldOnResource() throws CannotProvideStructureSchema {
    var modifiedResource = (ComplexStructureType) this.schemaProvider.provideSpecific(
        "DomainResource"
    );
    var actual = this.mapper.mapNewFields(
        modifiedResource,
        List.of("extension")
    );
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapNewFieldOnComplexType() throws CannotProvideStructureSchema {
    var modifiedResource = (ComplexStructureType) this.schemaProvider.provideSpecific(
        "CodeableConcept"
    );
    var actual = this.mapper.mapNewFields(
        modifiedResource,
        List.of("coding")
    );
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanMapNewInheritedFieldOnComplexType() throws CannotProvideStructureSchema {
    var modifiedResource = (ComplexStructureType) this.schemaProvider.provideSpecific(
        "Element"
    );
    var actual = this.mapper.mapNewFields(
        modifiedResource,
        List.of("extension")
    );
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }

  @Test
  void itCanDealWithContentReference() throws CannotProvideStructureSchema {
    var input = this.schemaProvider.provideSpecific("ValueSet");
    var actual = this.mapper.map((ResourceStructureType) input);
    this.thenObjectApproved(
        actual,
        RenderFeature.SORT_FIELDS,
        RenderFeature.RENDER_GETTERS
    );
  }
}