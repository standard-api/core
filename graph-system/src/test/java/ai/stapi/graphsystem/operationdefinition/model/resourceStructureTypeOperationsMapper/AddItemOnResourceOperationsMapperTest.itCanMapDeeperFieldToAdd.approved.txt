{
  "newParameters" : [ ],
  "operations" : [ {
    "id" : "AddTimeOfDayOnExampleResourceTypeExampleElementFieldRepeat",
    "code" : "AddTimeOfDayOnExampleResourceTypeExampleElementFieldRepeat",
    "description" : "Generated command for adding timeOfDay(time) on ExampleResourceType Aggregate",
    "instance" : true,
    "kind" : "operation",
    "name" : "Add TimeOfDay on ExampleResourceType.exampleElementField.repeat",
    "status" : "draft",
    "system" : false,
    "type" : false,
    "parameter" : [ {
      "max" : "*",
      "min" : 1,
      "name" : "timeOfDay",
      "type" : "time",
      "use" : "in",
      "referencedFrom" : [ {
        "source" : "ExampleResourceType.exampleElementField.repeat.timeOfDay"
      } ],
      "targetProfileReference" : [ ]
    } ],
    "resource" : [ "ExampleResourceType" ]
  }, {
    "id" : "AddEventOnExampleResourceTypeExampleElementField",
    "code" : "AddEventOnExampleResourceTypeExampleElementField",
    "description" : "Generated command for adding event(dateTime) on ExampleResourceType Aggregate",
    "instance" : true,
    "kind" : "operation",
    "name" : "Add Event on ExampleResourceType.exampleElementField",
    "status" : "draft",
    "system" : false,
    "type" : false,
    "parameter" : [ {
      "max" : "*",
      "min" : 1,
      "name" : "event",
      "type" : "dateTime",
      "use" : "in",
      "referencedFrom" : [ {
        "source" : "ExampleResourceType.exampleElementField.event"
      } ],
      "targetProfileReference" : [ ]
    } ],
    "resource" : [ "ExampleResourceType" ]
  }, {
    "id" : "AddDayOfWeekOnExampleResourceTypeExampleElementFieldRepeat",
    "code" : "AddDayOfWeekOnExampleResourceTypeExampleElementFieldRepeat",
    "description" : "Generated command for adding dayOfWeek(code) on ExampleResourceType Aggregate",
    "instance" : true,
    "kind" : "operation",
    "name" : "Add DayOfWeek on ExampleResourceType.exampleElementField.repeat",
    "status" : "draft",
    "system" : false,
    "type" : false,
    "parameter" : [ {
      "max" : "*",
      "min" : 1,
      "name" : "dayOfWeek",
      "type" : "code",
      "use" : "in",
      "referencedFrom" : [ {
        "source" : "ExampleResourceType.exampleElementField.repeat.dayOfWeek"
      } ],
      "targetProfileReference" : [ ]
    } ],
    "resource" : [ "ExampleResourceType" ]
  }, {
    "id" : "AddCodingOnExampleResourceTypeExampleElementFieldCode",
    "code" : "AddCodingOnExampleResourceTypeExampleElementFieldCode",
    "description" : "Generated command for adding coding(Coding) on ExampleResourceType Aggregate",
    "instance" : true,
    "kind" : "operation",
    "name" : "Add Coding on ExampleResourceType.exampleElementField.code",
    "status" : "draft",
    "system" : false,
    "type" : false,
    "parameter" : [ {
      "max" : "*",
      "min" : 1,
      "name" : "coding",
      "type" : "Coding",
      "use" : "in",
      "referencedFrom" : [ {
        "source" : "ExampleResourceType.exampleElementField.code.coding"
      } ],
      "targetProfileReference" : [ ]
    } ],
    "resource" : [ "ExampleResourceType" ]
  }, {
    "id" : "AddWhenOnExampleResourceTypeExampleElementFieldRepeat",
    "code" : "AddWhenOnExampleResourceTypeExampleElementFieldRepeat",
    "description" : "Generated command for adding when(code) on ExampleResourceType Aggregate",
    "instance" : true,
    "kind" : "operation",
    "name" : "Add When on ExampleResourceType.exampleElementField.repeat",
    "status" : "draft",
    "system" : false,
    "type" : false,
    "parameter" : [ {
      "max" : "*",
      "min" : 1,
      "name" : "when",
      "type" : "code",
      "use" : "in",
      "referencedFrom" : [ {
        "source" : "ExampleResourceType.exampleElementField.repeat.when"
      } ],
      "targetProfileReference" : [ ]
    } ],
    "resource" : [ "ExampleResourceType" ]
  } ]
}